library/alloc/src/collections/btree/node.rs

```diff
@@ -613,11 +613,11 @@ impl<'a, K: 'a, V: 'a> NodeRef<marker::Mut<'a>, K, V, marker::Leaf> {
     fn push_front(&mut self, key: K, val: V) {
         assert!(self.len() < CAPACITY);
 
+        self.as_leaf_mut().len += 1;
```

Comment:
I considered this solution, but... it's not great. Specifically, this means `self.keys_mut()` now returns a slice to memory that is not initialized. This is UB right now (even though I am in favor of not making this UB).

How bad would it be if `keys_mut`/`vals_mut` returned raw pointer slices instead of references? Or at least, if there also was a raw-pointer version for the insert case? (The reference version is still fine for all the other cases I think.)

Reply:
> memory that is not initialized

Great, so MaybeUninit memory has to be initialized, `MaybeUninit::uninit_array` is not ehough to do that? You excel pulling the rug from under all my beliefs!

> How bad would it be

Well, I sure have no idea. I can (hopefully) write, compile and test it, but I don't even know how to run Miri at the moment, and I thought the [CI failing](https://github.com/rust-lang/miri) status meant I wasn't supposed to know.

Reply:
> Great, so MaybeUninit memory has to be initialized, MaybeUninit::uninit_array is not enough to do that? You excel at pulling the rug from under all my beliefs!

The type of the involved function is
```rust
    fn keys_mut(&mut self) -> &mut [K]
```
So, there is no `MaybeUninit` here. Memory in a `MaybeUninit` indeed does not have to be initialized, but this one is not in a `MaybeUninit` (as far as the return type is concerned).

Reply:
Miri's CI will pass again once https://github.com/rust-lang/rust/pull/78446 lands. But you can always use the Miri from rustup; see https://rust-lang.github.io/rustup-components-history/ to find a nightly that comes with a working Miri.

However, to test if this fixes https://github.com/rust-lang/rust/issues/78477, https://github.com/rust-lang/miri/pull/1603 will have to land first (or you'll have to build Miri from that branch yourself).

Reply:
> So, there is no `MaybeUninit` here.

Right, I understand. I guess inserting into a slice of genuine values is a bad idea, because there is no way for them to know how to expand their footprint.

I do use Miri from rustup, but sometimes and right now it doesn't swallow the code in the Rust master branch. Until a day or two later, rustup finds an update and all is right again. I managed to build Miri in the past, but never figured out how to apply that in a x.py build.

Reply:
> Right, I understand. I guess inserting into a slice of genuine values is a bad idea, because there is no way for a slice to know how to expand its footprint.

Exactly. However, making those functions return a slice of `MaybeUninit` is probably the best solution (assuming this is not too hard on other callers). Then the slice no longer claims to be fully initialized. (However this means the function needs to be made `unsafe` as it relies on the caller not doing weird stuff like putting a `MaybeUninit::uninit` into one of the existing slice elements.)

> I do use Miri from rustup, but sometimes and right now it doesn't swallow the code in the Rust master branch. Until a day or two later, rustup finds an update and all is right again. I managed to build Miri in the past, but never figured out how to apply that in a x.py build.

Since you just need the standard library from an x.py build, something like this should work:
* Clone [this branch](https://github.com/RalfJung/miri/tree/track-raw)
* check rust-version file, and use that commit as your basis for the rustc PR
* in Miri, `XARGO_RUST_SRC=/path/to/rustc/library ./miri run tests/run-pass/btreemap.rs -Zmiri-track-raw-pointers`

Reply:
I get past this upheaval when establishing the slice using a safe `&mut keys[..len]`. But not using `slice::from_raw_parts_mut(keys.get_unchecked_mut(0), len)`. -Zmiri-track-pointer-tag says:

```
527 |             slice::from_raw_parts_mut(self.as_leaf_mut().vals.get_unchecked_mut(0), self.len())
    |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ created tag 14820
```

As if Miri ignores the slice function and the resulting `&mut [MaybeUninit<K>]` and clings onto the first &mut it saw at that address.

Reply:
Because Miri says that's the parent tag that does not have an appropriate item in the borrow stack.

This is the original error I get, using 17cc9b6256c95c31944591aec683884fead4e3b6: "Undefined Behavior: trying to reborrow for Unique at alloc1903+0x20, but parent tag <5115> does not have an appropriate item in the borrow stack". That stays the same for the code I parked here, with that `slice::from_raw_parts_mut(keys.get_unchecked_mut(0), len)`, except that the tag lowers to 5083.

Defining `keys_mut` as `&mut keys[..len]` (just for keys, not values and edges), that problem disappears, but obviously the same error occurs for values, with tag 14820.

It also disappears when defining `keys_mut` as:
```
let data = &mut self.as_leaf_mut().keys as *mut _ as *mut MaybeUninit<K>;
unsafe { slice::from_raw_parts_mut(data, self.len()) }
```

Though that's not the end of this: when I do the same for values and edges, there's another problem elsewhere.

Reply:
> Because Miri says that's the parent tag that does not have an appropriate item in the borrow stack.

But where does that error occur?

`self.as_leaf_mut().vals.get_unchecked_mut(0)` doesn't make much sense -- this is a reference *to the first element*, it must not be used to access later elements.

Reply:
To get a pointer that is usable for the entire slice, to `self.as_leaf_mut().as_mut_ptr().add(idx)`. (Something like `get_unchecked_mut(..len)` might also work.)

Reply:
> But where does that error occur?

Oh right, sorry, that was indeed in the same statement, and not way later in the `slice_insert` function.

Reply:
> Something like `get_unchecked_mut(0..len)`

indeed works too.